src/column/writer.rs

```diff
@@ -0,0 +1,881 @@
+        Encoding::RLE,
+        &self.rep_levels_sink[..],
+        self.descr.max_rep_level()
+      )?
+    } else {
+      vec![]
+    };
+
+    // Encoded length of repetition levels.
+    let rep_levels_byte_len = rep_levels.len();
+    // Encoded length of definition levels.
+    let def_levels_byte_len = def_levels.len();
+    // Uncompressed length in bytes, compressed length depends on writer version.
+    let uncompressed_size = rep_levels_byte_len + def_levels_byte_len + values.len();
+    // Select encoding based on current encoder and writer version (v1 or v2).
+    let encoding = if self.dict_encoder.is_some() {
+      self.props.dictionary_data_page_encoding()
+    } else {
+      self.encoder.encoding()
+    };
+
+    let data_page = match self.props.writer_version() {
+      WriterVersion::PARQUET_1_0 => {
+        // Process as data page v1.
+        rep_levels.extend_from_slice(&def_levels[..]);
+        rep_levels.extend_from_slice(values.data());
+        let buffer = match self.compressor {
+          Some(ref mut cmpr) => {
+            // TODO: figure out capacity for the compressed buffer
+            let mut compressed_buf = Vec::with_capacity(rep_levels.len() / 2);
+            cmpr.compress(&rep_levels[..], &mut compressed_buf)?;
+            compressed_buf
+          }
+          None => rep_levels
+        };
+
+        Page::DataPage {
+          buf: ByteBufferPtr::new(buffer),
+          num_values: self.num_buffered_values,
+          encoding: encoding,
+          def_level_encoding: Encoding::RLE,
+          rep_level_encoding: Encoding::RLE,
+          // TODO: process statistics
+          statistics: None
+        }
+      },
+      WriterVersion::PARQUET_2_0 => {
+        // Process as data page v2.
+        let mut compressed_values = Vec::with_capacity(values.len());
+        match self.compressor {
+          Some(ref mut cmpr) => cmpr.compress(values.data(), &mut compressed_values)?,
+          None => compressed_values.extend_from_slice(values.data())
+        }
+        rep_levels.extend_from_slice(&def_levels[..]);
+        rep_levels.extend_from_slice(&compressed_values[..]);
+        let buffer = rep_levels;
+
+        Page::DataPageV2 {
+          buf: ByteBufferPtr::new(buffer),
+          num_values: self.num_buffered_values,
+          encoding: encoding,
+          num_nulls: self.num_buffered_values - self.num_buffered_encoded_values,
+          num_rows: self.num_buffered_rows,
+          def_levels_byte_len: def_levels_byte_len as u32,
+          rep_levels_byte_len: rep_levels_byte_len as u32,
+          is_compressed: self.compressor.is_some(),
+          // TODO: process statistics
+          statistics: None
+        }
+      }
+    };
+
+    let compressed_page = CompressedPage::new(data_page, uncompressed_size);
+
+    // Check if we need to buffer data page or flush it to the sink directly.
+    if self.dict_encoder.is_some() {
+      self.data_pages.push_back(compressed_page);
```

Comment:
Nah, we cannot flush them, because we need to write dictionary page first. Data pages should be fairly small after dictionary encoding, and there are settings users can specify.